### ISSUE TITLE
[java] OutOfMemoryError on large project 2M LOC

### ISSUE THREAD

**Issue by nareshl119** [pmd/pmd#1379]
**Affects PMD Version:** 
6.3.0

**Description:**

We have implemented PDM custom rules and  we are running the PMD custom rules against a large project (2M LOC) that using gradle for build results in OutOfMemoryError during PMD run. I could not find any way to specify to PMD  the maximum amount of memory it may consume in the build.gradle or publish.gradle file.

Please let me know how to resolve the above problem and what would be PMD scan time for larger projects like 2M LOC

![image](https://user-images.githubusercontent.com/39321364/46814762-66733b00-cd47-11e8-8d8b-32f4d18d7e29.png)

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

Gradle

java.lang.OutOfMemoryError: GC overhead limit exceeded

Thank you.

**Comment by jsotuyod** [comment]
@nareshl119 gradle runs PMD in-process, so setting the memory for Gradle itself works. This is usually done from within your `gradle.properties` file using:

```groovy
// set the required memory, 4G in this example
org.gradle.jvmargs=-Xmx4G
```

Non the less, it would be most interesting for us to investigate further where the memory is being eaten up. I can think of only 2 reasons:

1. the project has too many violations, as violation s are kept in memory until the rendering of the report is done at the end.
2. your custom rules are keeping node references around which prevent the ASTs from being garbage collected.

If you run PMD without your custom rules, do you hit the same issue?

**Comment by nareshl119** [comment]
Thank you for the quick response Martino.

I have increased the memory size as you suggested. But, still getting the same exception.

I ran PMD without custom rules, didn't the same issue.

Could you please investigate more into this and let me know.

Thank you,
Naresh

**Comment by jsotuyod** [comment]
@nareshl119 well if without the custom rules you don't get the exception, chances are high the problem lies within how you wrote the rules... are these XPath rules? or are they written in Java?

**Comment by nareshl119** [comment]
The rules are written in java

**Comment by jsotuyod** [comment]
@nareshl119 this looks almost definitively as a problem with your rules.

Is this an Open Source project or is it private?

My advice is to make sure your rules are keeping as little state as possible (ideally none!) and that such states has no reference to any `Node`s.

**Comment by nareshl119** [comment]
@jsotuyod I don't think there is a problem with the rules because I ran my rules against different projects which are 45000 LOC, running perfectly without any exceptions and I am getting violation report as well.

For the current project which is  2M LOC, I have excluded some of the folders from the root directory of the project to not to run the PMD rules against those folders then I am getting the violation report for remaining folder files without any exception. 

When I ran the PMD rules for the all the folders(root directory of the project) then getting the exception.

It's not an Open source project.

**Comment by jsotuyod** [comment]
@nareshl119 hard to tell, there are lots of variables in place. PMD itself is ~250k SLoC and on my machine a clean non-incremental analysis with type resolution data (your exact same scenario given you are using Gradle) takes ~100 secs.

You may want to play with the memory a bit. That 4G was an example and shouldn't be taken as a reference.

If this is a large corporate repository we can't share publicly, please consider reaching out privately to us for better support. My email is publicly available on my profile.

**Comment by nareshl119** [comment]
@jsotuyod Do you know how the scan time increases when new rules are added or the number files to be scanned increase? E.g. if it's linearly, then when number of rules is doubled then it takes twice as long?

**Comment by jsotuyod** [comment]
@nareshl119 from a codebase point of view, it's linear. However, as you are seeing a OOM, you are probably having a lot of time spent in the GC before you hit that error, and that is probably not linear.

PMD allows for multithreaded analysis, but Gradle doesn't use it; as it allocates a single worker thread per task.

**Comment by nareshl119** [comment]
@jsotuyod I am just curious to know how the PMD scan time will be in general. If it is linear then when number of rules is doubled then it takes twice as long?

**Comment by nareshl119** [comment]
@jsotuyod As we know PMD supports Incremental Analysis - 

https://pmd.github.io/pmd-6.3.0/pmd_userdocs_getting_started.html#enabling-incremental-analysis

Could you please let me know hot to define the location to store the cache.

**Comment by jsotuyod** [comment]
@nareshl119 not exactly.

Here is a sample output from PMD's benchmark on PMD itself (multithreaded run, all measurement are thread-based except for the Wall Clock Time):

```
--------------------------------------------<<< Rule >>>--------------------------------------------
Label                                              Time (secs) Self Time (secs)  # Calls     Counter

CloneMethodMustImplementCloneable                       0.0623           0.0623    2,586
UnnecessaryCaseChange                                   0.1798           0.1798    2,586
ImportFromSamePackage                                   0.1879           0.1879    2,586
IdempotentOperations                                    0.1915           0.1915    2,586
SuspiciousHashcodeMethodName                            0.1968           0.1968    2,586
SuspiciousOctalEscape                                   0.2283           0.2283    2,586
CompareObjectsWithEquals                                0.2293           0.2293    2,586
AvoidFieldNameMatchingMethodName                        0.2348           0.2348    2,586
SingletonClassReturningNewInstance                      0.2359           0.2359    2,586
AvoidFieldNameMatchingTypeName                          0.2370           0.2370    2,586
AvoidCatchingThrowable                                  0.2492           0.2492    2,586
AssignmentToNonFinalStatic                              0.2773           0.2773    2,586
DontImportSun                                           0.2792           0.2792    2,586
AvoidCallingFinalize                                    0.2973           0.2973    2,586
UnnecessaryConversionTemporary                          0.3274           0.3274    2,586
CheckSkipResult                                         0.3508           0.3508    2,586
UselessOperationOnImmutable                             0.3756           0.3756    2,586
BrokenNullCheck                                         0.3908           0.3908    2,586
AvoidUsingOctalValues                                   0.3997           0.3997    2,586
BeanMembersShouldSerialize                              0.4401           0.4401    2,586
SingleMethodSingleton                                   0.4595           0.4595    2,586
OverrideBothEqualsAndHashcode                           0.4634           0.4634    2,586
MethodWithSameNameAsEnclosingClass                      0.4902           0.4902    2,586
AvoidDuplicateLiterals                                  0.4973           0.4973    2,586
CloseResource                                           0.5186           0.5186    2,586
AssignmentInOperand                                     0.6846           0.6846    2,586
AvoidLiteralsInIfCondition                              4.4835           4.4835    2,586
ConstructorCallsOverridableMethod                       4.8464           4.8464    2,586
TestClassWithoutTestCases                               6.7422           6.7422    2,586
UnusedNullCheckInEquals                                 7.1571           7.1571    2,586
DataflowAnomalyAnalysis                                 9.9350           9.9350    2,586
MoreThanOneLogger                                      14.4372          14.4372    2,586

Total Rule                                             56.0861          56.0861

---------------------------------------<<< Rulechain Rule >>>---------------------------------------
Label                                              Time (secs) Self Time (secs)  # Calls     Counter

EmptyInitializer                                        0.0052           0.0052    2,586          86
AvoidMultipleUnaryOperators                             0.0052           0.0052    2,586       4,139
EmptySynchronizedBlock                                  0.0054           0.0054    2,586           5
EmptyFinallyBlock                                       0.0060           0.0060    2,586       1,094
NonStaticInitializer                                    0.0083           0.0083    2,586          86
DontUseFloatTypeForLoopIndices                          0.0133           0.0133    2,586       1,621
NullAssignment                                          0.0138           0.0138    2,586       4,635
EmptyTryBlock                                           0.0196           0.0196    2,586       1,945
AvoidBranchingStatementAsLastInLoop                     0.0234           0.0234    2,586      29,702
EmptyCatchBlock                                         0.0249           0.0249    2,586       1,837
JumbledIncrementer                                      0.0250           0.0250    2,586       1,621
AvoidCatchingNPE                                        0.0260           0.0260    2,586       1,837
DoNotThrowExceptionInFinally                            0.0450           0.0450    2,586       1,094
DoNotExtendJavaLangThrowable                            0.0502           0.0502    2,586       3,248
EmptySwitchStatements                                   0.0508           0.0508    2,586       1,955
EmptyStatementNotInLoop                                 0.0672           0.0672    2,586         656
ReturnFromFinallyBlock                                  0.0712           0.0712    2,586       1,094
EmptyWhileStmt                                          0.0725           0.0725    2,586         821
AvoidDecimalLiteralsInBigDecimalConstructor             0.0768           0.0768    2,586       7,290
SimpleDateFormatNeedsLocale                             0.0868           0.0868    2,586       7,290
ProperCloneImplementation                               0.0932           0.0932    2,586      19,847
StaticEJBFieldShouldBeFinal                             0.1040           0.1040    2,586       3,248
ClassCastExceptionWithToArray                           0.1077           0.1077    2,586       4,750
EmptyIfStmt                                             0.1190           0.1190    2,586      20,217
FinalizeOnlyCallsSuperFinalize                          0.1265           0.1265    2,586      19,847
AvoidInstanceofChecksInCatchClause                      0.1318           0.1318    2,586       1,837
BadComparison                                           0.1327           0.1327    2,586       9,472
FinalizeOverloaded                                      0.1331           0.1331    2,586      19,847
StringBufferInstantiationWithChar                       0.1390           0.1390    2,586       7,290
InvalidSlf4jMessageFormat                               0.1495           0.1495    2,586     223,959
UseEqualsToCompareStrings                               0.1527           0.1527    2,586       9,472
FinalizeDoesNotCallSuperFinalize                        0.1725           0.1725    2,586      19,847
LoggerIsNotStaticFinal                                  0.1744           0.1744    2,586      20,989
AvoidLosingExceptionInformation                         0.1797           0.1797    2,586       1,837
CloneMethodMustBePublic                                 0.2227           0.2227    2,586      19,847
MissingSerialVersionUID                                 0.2235           0.2235    2,586       3,248
MissingStaticMethodInNonInstantiatableClass             0.2269           0.2269    2,586       3,248
CloneThrowsCloneNotSupportedException                   0.2385           0.2385    2,586      19,847
EmptyFinalizer                                          0.2457           0.2457    2,586      19,847
UnconditionalIfStatement                                0.2548           0.2548    2,586      20,217
SuspiciousEqualsMethodName                              0.2556           0.2556    2,586      19,847
ReturnEmptyArrayRatherThanNull                          0.2741           0.2741    2,586      19,847
ProperLogger                                            0.2800           0.2800    2,586      29,425
CloneMethodReturnTypeMustMatchClassName                 0.3572           0.3572    2,586      19,847
EmptyStatementBlock                                     0.3833           0.3833    2,586     109,415
JUnitStaticSuite                                        0.3864           0.3864    2,586      19,847
CallSuperLast                                           0.4174           0.4174    2,586      19,847
AvoidAssertAsIdentifier                                 0.4193           0.4193    2,586      44,257
AvoidEnumAsIdentifier                                   0.4656           0.4656    2,586      44,257
FinalizeShouldBeProtected                               0.5321           0.5321    2,586      19,847
NonCaseLabelInSwitchStatement                           0.5557           0.5557    2,586       1,955
UseCorrectExceptionLogging                              0.5955           0.5955    2,586       1,837
UseProperClassLoader                                    0.6153           0.6153    2,586      90,699
InstantiationToGetClass                                 0.6214           0.6214    2,586      90,699
UnnecessaryBooleanAssertion                             0.6396           0.6396    2,586      47,503
CallSuperFirst                                          0.7276           0.7276    2,586      19,847
MisplacedNullCheck                                      1.1961           1.1961    2,586     254,654
DoNotHardCodeSDCard                                     1.3321           1.3321    2,586     113,251
JUnitSpelling                                           1.6269           1.6269    2,586      19,847
DoNotCallSystemExit                                     2.2304           2.2304    2,586     223,959
DoNotCallGarbageCollectionExplicitly                    2.7201           2.7201    2,586     223,959
EqualsNull                                              3.3929           3.3929    2,586     342,262
UseLocaleWithCaseConversions                            4.5755           4.5755    2,586     342,262
MissingBreakInSwitch                                    4.9960           4.9960    2,586       1,955
AvoidAccessibilityAlteration                            6.5773           6.5773    2,586     342,262

Total Rulechain Rule                                   40.1983          40.1983

------------------------------------------<<< Summary  >>>------------------------------------------
Label                                              Time (secs) Self Time (secs)  # Calls     Counter

Rule                                                   56.2645           0.1784    2,586
Rulechain Rule                                         40.3256           0.1274    2,586
Collect Files                                           0.4893           0.4893        1
Load Rules                                              0.2708           0.2708        1
Parser                                                 16.7500          16.7500    2,598
Qualified Name Resolution                               1.9929           1.9929    2,586
Symbol Table                                           62.3645          62.3645    2,586
Dfa                                                     5.5332           5.5332    2,586
Type Resolution                                        92.1752          92.1752    2,586
Rulechain Visit                                         0.8011           0.8011    2,586
Reporting                                               0.0837           0.0837    2,601
File Processing                                        77.7789          77.7024        1
Unaccounted                                             1.6368           1.6368

-------------------------------------------<<< Total  >>>-------------------------------------------
Label                                              Time (secs) Self Time (secs)  # Calls     Counter

Wall Clock Time                                        78.6310
```

As seen here:

* there is great variance from rule to rule, depending on how heavyweight they are. Some are blazing fast, others have room for improvement.
* most of the time is not spent on the rule, but on:
  * ~17s are spent parsing Java files
  * ~62s are spent resolving symbols
  * ~92s are spent resolving types

Those 3 account for more than 50% of PMD processing.

**Comment by jsotuyod** [comment]
> @jsotuyod As we know PMD supports Incremental Analysis -
> 
> https://pmd.github.io/pmd-6.3.0/pmd_userdocs_getting_started.html#enabling-incremental-analysis
> 
> Could you please let me know hot to define the location to store the cache.

You are using Gradle, which doesn't currently support the incremental analysis feature. Support for it is a frequent request https://github.com/gradle/gradle/pull/3125

**Comment by regrog** [comment]
> PMD allows for multithreaded analysis, but Gradle doesn't use it; as it allocates a single worker thread per task.

Hi,
I'm facing OOM errors during PMD (6.18) analysis using Gradle (5.6.2).
About multithread in Gradle, do you have any news?
Do you think it's something that can be achieved in Gradle? Like they did with incremental analysis?
Thanks

**Comment by jsotuyod** [comment]
I'm closing this as stale due to inactivity / lack of enough data to dig deeper, with all current data pointing at issues with custom Java rules.

Feel free to log a new issue with any further data if this is still an issue under current versions.
